### Part I: Indexing

__Exercise 1.1 (10 points).__ Give three examples of indexing a data frame with `[ ]`, `.loc[ ]`, and `.iloc[ ]`, respectively. Explain how each of these indexing methods is different.

In [36]:
import pandas as pd
df = pd.DataFrame({"x": [1, 2, 3], "y": ["a", "b", "c"]}) #generate a data frame first
df

x  y
0  1  a
1  2  b
2  3  c

In [37]:
display(df["x"])
display(df.x[1])
df [['x','y'] ]  

0    1
1    2
2    3
Name: x, dtype: int64

2

x  y
0  1  a
1  2  b
2  3  c

**[ ] can be used to select object by position, name, or condition.**

**First example shows we select the column which name is "x".**

**Second example shows we select the second object of column x.**

**Third example shows we select two columns which are x and y.**

In [5]:
display(df.loc[:, "x"])
display(df.loc[df.x ==1])
display(df.loc[(df.x > 0) &(df.y!= 'c')])

0    1
1    2
2    3
Name: x, dtype: int64

x  y
0  1  a

x  y
0  1  a
1  2  b

**.loc[ ] can be used to select object by name or condition.**

**First example shows we select all the rows for column x.**

**Second example shows we select the row when x is 1.**

**Third example shows we select the row when x is greater than 0 and y not equal to c.**

In [6]:
display(df.iloc[1,1])
display(df.iloc[1:3,1])
display(df.iloc[:,1])

'b'

1    b
2    c
Name: y, dtype: object

0    a
1    b
2    c
Name: y, dtype: object

**.iloc[ ] can be used to select object by position.**

**First example shows we select the object at second row and second column.**

**Second example shows we select the second and third row (index is 1 and 2) at the second column (y).**

**Third example shows we select all the rows at the second column (y).**

__Exercise 1.2 (10 points).__ What do negative indexes (as in `x[-1]`) do in Python? Create and try three examples of negative indexes for lists, NumPy arrays, and data frames, respectively. Then explain what you think negative indexes do. Confirm your explanation by linking to a relevant page in the Python, NumPy, or Pandas documentation.

In [93]:
x=[1,2,3,4,5]
print(x)
x[-1]

[1, 2, 3, 4, 5]


5

In [94]:
x = np.array([[1, 2, 3], [4, 5, 6]])
print(x)
x[-1]

[[1 2 3]
 [4 5 6]]


array([4, 5, 6])

In [95]:
df = pd.DataFrame({"x": [1, 2, 3], "y": ["a", "b", "c"]})
display(df)
display(df[-1:])
display(df[:-1])

x  y
0  1  a
1  2  b
2  3  c

x  y
2  3  c

x  y
0  1  a
1  2  b

**In the list example, x[-1] only select the last element in the list.**

**In the numpy array example, x[-1] select the last dimemsion of the array.**

**If we change [-1] to [:-1], it will be totally opposite. It will choose all but the last element.**

**In the dataframe example, df[-1:] select the last row for all columns.**

__Exercise 1.3 (10 points).__ Give an example and explain Pandas' data alignment (or index alignment) feature.

In [137]:
df1 = pd.DataFrame([['Ana',25,120], ['Bob',35,180]], columns=['Name', 'Age', 'Weight'])
df1 = df1.set_index("Name")
display(df1)
df2 = pd.DataFrame([['Bob',185,'twice'], ['Ana',160,'once']], columns= ['Name', 'Height', 'Exercise frequency per week'])
df2 = df2.set_index("Name")
df2

Age  Weight
Name             
Ana    25     120
Bob    35     180

Height Exercise frequency per week
Name                                    
Bob      185                       twice
Ana      160                        once

In [138]:
full1_df = pd.merge(left=df1, right=df2, left_index=True, right_index=True)
full1_df

Age  Weight  Height Exercise frequency per week
Name                                                 
Ana    25     120     160                        once
Bob    35     180     185                       twice

In [132]:
df1 = pd.DataFrame([['Ana',25,120], ['Bob',35,180]], columns=['Name', 'Age', 'Weight'])
df1 = df1.set_index("Name")
display(df1)
df2 = pd.DataFrame([['Bob',185,'twice']], columns= ['Name', 'Height', 'Exercise frequency per week'])
df2 = df2.set_index("Name")
display(df2)

Age  Weight
Name             
Ana    25     120
Bob    35     180

Height Exercise frequency per week
Name                                    
Bob      185                       twice

In [133]:
full2_df = pd.merge(left=df1, right=df2, left_index=True, right_index=True)
full2_df

Age  Weight  Height Exercise frequency per week
Name                                                 
Bob    35     180     185                       twice

**In this example, we have to set the index as Name, otherwise it will align the data frame based on the default index. Also, the number of rows should be the same, otherwise it will only align the rows with same index.**

In [291]:
x = pd.Series([10, 20, 30], index = ["a", "b", "c"])
y = pd.Series([10, 20, 30, 10], index = ["b", "a", "d", "c"])
display(x)
display(y)
x+y

a    10
b    20
c    30
dtype: int64

b    10
a    20
d    30
c    10
dtype: int64

a    30.0
b    30.0
c    40.0
d     NaN
dtype: float64

**In this example, elements are automatically aligned by index. For index a, 10+20=30. They are not alighned by position.**

**The feature of Pandas' data alignment is that it can merge the two dataframes based on the index. It's a powerful tool when you have multiple variables in different dataset measured from same group of objects. It can merge them automaticlly. But if we want to have operation based on position, we can reset the index to add another column as index. Then the data can align based on position and order.**

__Exercise 1.4 (10 points).__ Explain the difference between the similarly-named data frame methods `.reindex()` and `.reset_index()`. Give two examples to show what each method respectively does.

How might these methods be useful when combined with Pandas' data alignment feature?

*Hint: Besides the Pandas documentation, `.reindex()` is explained in Python for Data Analysis 5.2, and `.reset_index()` is explained [here](https://jakevdp.github.io/PythonDataScienceHandbook/03.05-hierarchical-indexing.html#Index-setting-and-resetting).*

In [162]:
df1 = pd.DataFrame([['Ana',25,120], ['Bob',35,180],['Charlie',40,190]], columns=['Name', 'Age', 'Weight'], index=['b','c','a'])
display(df1)
df1.reindex(['a', 'b', 'c']) #example 1

Name  Age  Weight
b      Ana   25     120
c      Bob   35     180
a  Charlie   40     190

Name  Age  Weight
a  Charlie   40     190
b      Ana   25     120
c      Bob   35     180

In [165]:
df2 = pd.DataFrame(['Ana', 'Bob', 'Charlie'], index=[0, 2, 4])
display(df2)
df2.reindex(range(6), method='ffill')

0
0      Ana
2      Bob
4  Charlie

0
0      Ana
1      Ana
2      Bob
3      Bob
4  Charlie
5  Charlie

**.reindex() method allows us to rearrage the order of the data based on the index. In the first example, this method helps us to order the data from index a to c. In the second example, it allows us change the index from 0, 2, 4 to 1, 2, 3, 4, 5 and fill the missing value.**

In [177]:
df1 = pd.DataFrame([['Ana',25,120], ['Bob',35,180]], columns=['Name', 'Age', 'Weight'], index=['Patient-1', 'Patient-2'])
display(df1)
df1.reset_index()

Name  Age  Weight
Patient-1  Ana   25     120
Patient-2  Bob   35     180

index Name  Age  Weight
0  Patient-1  Ana   25     120
1  Patient-2  Bob   35     180

In [178]:
df1.reset_index(drop=True)

Name  Age  Weight
0  Ana   25     120
1  Bob   35     180

**When we use reset_index() method, a new index was added to the first column and keep the previous column as before. In the first example, we can the it added a new column as new index.
If we don't want the old index, we can use the drop=T to get rid of the old index. In the second example, we can the old index was removed but the data has a new index column.**

**Both methods are able to create new index, which is very useful to align data. Because pandas data alignment requies two dataframes with the same row and/or column configuration based on the index. After we make the two data frame have the same index, pandas data alignment can merge them automatically based on the common index.**


### Part 2: Read in the 'county_age_dist.csv' data and answer the next four questions.

__Exercise 2.1 (5 points).__ Cast the 'fips' column as a string such that it is a length 5 string of digits that starts with 0 if it is currently of length 4.  Yolo county for example is '06113'.  This is the age distribution of each county in the US (with territories), you may notice that the total population is unusually large, please ignore this.

In [9]:
import numpy as np
import pandas as pd
age = pd.read_csv("county_age_dist.csv")
# age.head()
age.dtypes #check types of each feature 

fips     int64
0-17     int64
18-24    int64
25-34    int64
35-44    int64
45-54    int64
55-64    int64
65-74    int64
75-84    int64
85+      int64
dtype: object

In [10]:
age = age.astype({"fips": str}) # change fips to string
age.dtypes #check types of each feature 
age['fips']= age['fips'].str.zfill(5) #find str.zfill at
# https://pandas.pydata.org/docs/reference/api/pandas.Series.str.zfill.html.
# to make the string to 5 digits
age.head()

fips   0-17  18-24  25-34  35-44  45-54  55-64  65-74  75-84   85+
0  01001  25941  11422  12315  13828  14000  12697   9594   5430  1945
1  01003  86587  37568  44133  46730  49675  52405  43252  23262  8854
2  01005  11057   6162   6603   5907   6490   6377   5255   2795  1074
3  01007   9671   5241   5788   5472   6707   5563   4270   2555   638
4  01009  25671  11360  12635  13570  14737  14123  12106   6560  2022

__Exercise 2.2 (5 points).__ Create a population column which is the sum of the age counts for all age bins.  For each age bin create a column which is proportion of the age bin in the county populations, so for example '0-17_rat' is the ratio of '0-17' to the whole population for the county.  Print out the FIPS code for the county with the largest proportion of 0-17 year olds? What is the FIPS code for the county with the largest proportion of 85+ year olds?

In [11]:
age['population'] =age.drop('fips', axis=1).sum(axis=1) #drop the first column and sum the rest of column
# similar code found at https://stackoverflow.com/questions/44003371/all-row-sum-with-pandas-except-one.
# calculate the sum for each row
age['0-17_rat']= age['0-17'] / age['population']  # add new column as the proportion for each age range
age['18-24_rat']= age['18-24'] / age['population'] 
age['25-34_rat']= age['25-34'] / age['population'] 
age['35-44_rat']= age['35-44'] / age['population'] 
age['45-54_rat']= age['45-54'] / age['population'] 
age['55-64_rat']= age['55-64'] / age['population'] 
age['65-74_rat']= age['65-74'] / age['population'] 
age['75-84_rat']= age['75-84'] / age['population'] 
age['85+_rat']= age['85+'] / age['population'] 
age.head()

fips   0-17  18-24  25-34  35-44  45-54  55-64  65-74  75-84   85+  \
0  01001  25941  11422  12315  13828  14000  12697   9594   5430  1945   
1  01003  86587  37568  44133  46730  49675  52405  43252  23262  8854   
2  01005  11057   6162   6603   5907   6490   6377   5255   2795  1074   
3  01007   9671   5241   5788   5472   6707   5563   4270   2555   638   
4  01009  25671  11360  12635  13570  14737  14123  12106   6560  2022   

   population  0-17_rat  18-24_rat  25-34_rat  35-44_rat  45-54_rat  \
0      107172  0.242050   0.106576   0.114909   0.129026   0.130631   
1      392466  0.220623   0.095723   0.112451   0.119068   0.126571   
2       51720  0.213786   0.119142   0.127668   0.114211   0.125483   
3       45905  0.210674   0.114171   0.126086   0.119203   0.146106   
4      112784  0.227612   0.100724   0.112028   0.120318   0.130666   

   55-64_rat  65-74_rat  75-84_rat   85+_rat  
0   0.118473   0.089520   0.050666  0.018148  
1   0.133527   0.110206   0.059271  0.022560  
2   0.123299   0.101605   0.054041  0.020766  
3   0.121185   0.093018   0.055658  0.013898  
4   0.125222   0.107338   0.058164  0.017928

In [12]:
print(age.sort_values(by=['0-17_rat'],ascending=False).iloc[0,0]) 
#sort the data (large-small) based on the 0-17 age ratio and then output is the fips of the max
print(age.sort_values(by=['85+_rat'],ascending=False).iloc[0,0])
#sort the data (large-small) based on the 85+ age ratio and then output is the fips of the max
#https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html .sort_value

02158
31091


**02158 is the fips of county with the largest proportion of 0-17 year olds. 31091 is the fips of county with the largest proportion of 85+ year olds.**

__Exercise 2.3 (5 points).__ What is the overall proportion of each age bin within the entire US.  You need to find the total number of each age bin across all counties and take the ratio with the total population.

In [13]:
a1 = sum(age['population']) #calculate population for entire US   
pd.DataFrame(data= age.sum()/a1, columns=['Proportion'])[1:10] #calculate proportion

Proportion
0-17     0.224884
18-24    0.117452
25-34    0.128200
35-44    0.119159
45-54    0.124014
55-64    0.123719
65-74    0.089391
75-84    0.050112
85+      0.023068

__Exercise 2.4 (10 points).__ Create a new column that contains the mode of the age bins (the age bin with the largest proportion in the county).  Print out the frequencies of the modes different possible values, so that the frequency of '0-17' is the proportion of the counties that have '0-17' as the largest age bin for that county.

In [14]:
y = age.iloc[:, 11:]
#y = age.select_dtypes(include = ["float64"]).drop('population',axis=1) #select all ratio column
y

0-17_rat  18-24_rat  25-34_rat  35-44_rat  45-54_rat  55-64_rat  \
0     0.242050   0.106576   0.114909   0.129026   0.130631   0.118473   
1     0.220623   0.095723   0.112451   0.119068   0.126571   0.133527   
2     0.213786   0.119142   0.127668   0.114211   0.125483   0.123299   
3     0.210674   0.114171   0.126086   0.119203   0.146106   0.121185   
4     0.227612   0.100724   0.112028   0.120318   0.130666   0.125222   
...        ...        ...        ...        ...        ...        ...   
3215  0.222862   0.119243   0.122765   0.116876   0.123953   0.119202   
3216  0.217616   0.125622   0.096370   0.108531   0.114981   0.136908   
3217  0.234818   0.133015   0.114193   0.110015   0.119715   0.128816   
3218  0.211635   0.118870   0.111495   0.121162   0.125759   0.129144   
3219  0.213282   0.110158   0.115657   0.116087   0.123096   0.131001   

      65-74_rat  75-84_rat   85+_rat  
0      0.089520   0.050666  0.018148  
1      0.110206   0.059271  0.022560  
2      0.101605   0.054041  0.020766  
3      0.093018   0.055658  0.013898  
4      0.107338   0.058164  0.017928  
...         ...        ...       ...  
3215   0.098397   0.050642  0.026059  
3216   0.108900   0.069007  0.022066  
3217   0.095785   0.045876  0.017767  
3218   0.104395   0.049840  0.027699  
3219   0.105211   0.056711  0.028798  

[3220 rows x 9 columns]

In [15]:
age['Mode'] = y.idxmax(axis = 1) #find the column index of the max for each row
# and use the index to generate new column
#https://www.geeksforgeeks.org/find-maximum-values-position-in-columns-and-rows-of-a-dataframe-in-pandas/
age

fips   0-17  18-24  25-34  35-44  45-54  55-64  65-74  75-84   85+  \
0     01001  25941  11422  12315  13828  14000  12697   9594   5430  1945   
1     01003  86587  37568  44133  46730  49675  52405  43252  23262  8854   
2     01005  11057   6162   6603   5907   6490   6377   5255   2795  1074   
3     01007   9671   5241   5788   5472   6707   5563   4270   2555   638   
4     01009  25671  11360  12635  13570  14737  14123  12106   6560  2022   
...     ...    ...    ...    ...    ...    ...    ...    ...    ...   ...   
3215  72145  27016  14455  14882  14168  15026  14450  11928   6139  3159   
3216  72147   4724   2727   2092   2356   2496   2972   2364   1498   479   
3217  72149  11353   6431   5521   5319   5788   6228   4631   2218   859   
3218  72151  16068   9025   8465   9199   9548   9805   7926   3784  2103   
3219  72153  17375   8974   9422   9457  10028  10672   8571   4620  2346   

      ...  0-17_rat  18-24_rat  25-34_rat  35-44_rat  45-54_rat  55-64_rat  \
0     ...  0.242050   0.106576   0.114909   0.129026   0.130631   0.118473   
1     ...  0.220623   0.095723   0.112451   0.119068   0.126571   0.133527   
2     ...  0.213786   0.119142   0.127668   0.114211   0.125483   0.123299   
3     ...  0.210674   0.114171   0.126086   0.119203   0.146106   0.121185   
4     ...  0.227612   0.100724   0.112028   0.120318   0.130666   0.125222   
...   ...       ...        ...        ...        ...        ...        ...   
3215  ...  0.222862   0.119243   0.122765   0.116876   0.123953   0.119202   
3216  ...  0.217616   0.125622   0.096370   0.108531   0.114981   0.136908   
3217  ...  0.234818   0.133015   0.114193   0.110015   0.119715   0.128816   
3218  ...  0.211635   0.118870   0.111495   0.121162   0.125759   0.129144   
3219  ...  0.213282   0.110158   0.115657   0.116087   0.123096   0.131001   

      65-74_rat  75-84_rat   85+_rat      Mode  
0      0.089520   0.050666  0.018148  0-17_rat  
1      0.110206   0.059271  0.022560  0-17_rat  
2      0.101605   0.054041  0.020766  0-17_rat  
3      0.093018   0.055658  0.013898  0-17_rat  
4      0.107338   0.058164  0.017928  0-17_rat  
...         ...        ...       ...       ...  
3215   0.098397   0.050642  0.026059  0-17_rat  
3216   0.108900   0.069007  0.022066  0-17_rat  
3217   0.095785   0.045876  0.017767  0-17_rat  
3218   0.104395   0.049840  0.027699  0-17_rat  
3219   0.105211   0.056711  0.028798  0-17_rat  

[3220 rows x 21 columns]

**The table above shows the mode column. The table below is the frequency of mode.**

In [16]:
tab = pd.crosstab(index=age["Mode"], columns='count')
tab
#tab/tab.sum()

col_0      count
Mode            
0-17_rat    3057
18-24_rat     70
25-34_rat     15
35-44_rat      2
45-54_rat      3
55-64_rat     58
65-74_rat     15

### Part 3. Read in the 'Read in the 'county_fips_master.csv'' data and answer the next two questions

__Exercise 3.1 (5 points).__ Join this dataset to the county age dataframe constructed above such that there is one row for every county in the new dataframe from county_fips_master. How many rows are there in the new dataframe? Are there any missing entries? If so how many rows have missingness?

In [28]:
master = pd.read_csv("county_fips_master.csv", sep=",",encoding="latin-1") #because the data is not utf8
#https://stackoverflow.com/questions/12349786/encoding-latin1-to-utf8-fails
master

fips        county_name state_abbr state_name             long_name  \
0      1001     Autauga County         AL    Alabama     Autauga County AL   
1      1003     Baldwin County         AL    Alabama     Baldwin County AL   
2      1005     Barbour County         AL    Alabama     Barbour County AL   
3      1007        Bibb County         AL    Alabama        Bibb County AL   
4      1009      Blount County         AL    Alabama      Blount County AL   
...     ...                ...        ...        ...                   ...   
3141  56037  Sweetwater County         WY    Wyoming  Sweetwater County WY   
3142  56039       Teton County         WY    Wyoming       Teton County WY   
3143  56041       Uinta County         WY    Wyoming       Uinta County WY   
3144  56043    Washakie County         WY    Wyoming    Washakie County WY   
3145  56045      Weston County         WY    Wyoming      Weston County WY   

      sumlev  region  division  state  county  crosswalk region_name  \
0       50.0     3.0       6.0    1.0     1.0    3-6-1-1       South   
1       50.0     3.0       6.0    1.0     3.0    3-6-1-3       South   
2       50.0     3.0       6.0    1.0     5.0    3-6-1-5       South   
3       50.0     3.0       6.0    1.0     7.0    3-6-1-7       South   
4       50.0     3.0       6.0    1.0     9.0    3-6-1-9       South   
...      ...     ...       ...    ...     ...        ...         ...   
3141    50.0     4.0       8.0   56.0    37.0  4-8-56-37        West   
3142    50.0     4.0       8.0   56.0    39.0  4-8-56-39        West   
3143    50.0     4.0       8.0   56.0    41.0  4-8-56-41        West   
3144    50.0     4.0       8.0   56.0    43.0  4-8-56-43        West   
3145    50.0     4.0       8.0   56.0    45.0  4-8-56-45        West   

           division_name  
0     East South Central  
1     East South Central  
2     East South Central  
3     East South Central  
4     East South Central  
...                  ...  
3141            Mountain  
3142            Mountain  
3143            Mountain  
3144            Mountain  
3145            Mountain  

[3146 rows x 13 columns]

In [29]:
master = master.astype({"fips": str}) # change fips to string
master.dtypes #check types of each feature 
master['fips']= master['fips'].str.zfill(5) #find str.zfill at
# https://pandas.pydata.org/docs/reference/api/pandas.Series.str.zfill.html.
# to make the string to 5 digits
master

fips        county_name state_abbr state_name             long_name  \
0     01001     Autauga County         AL    Alabama     Autauga County AL   
1     01003     Baldwin County         AL    Alabama     Baldwin County AL   
2     01005     Barbour County         AL    Alabama     Barbour County AL   
3     01007        Bibb County         AL    Alabama        Bibb County AL   
4     01009      Blount County         AL    Alabama      Blount County AL   
...     ...                ...        ...        ...                   ...   
3141  56037  Sweetwater County         WY    Wyoming  Sweetwater County WY   
3142  56039       Teton County         WY    Wyoming       Teton County WY   
3143  56041       Uinta County         WY    Wyoming       Uinta County WY   
3144  56043    Washakie County         WY    Wyoming    Washakie County WY   
3145  56045      Weston County         WY    Wyoming      Weston County WY   

      sumlev  region  division  state  county  crosswalk region_name  \
0       50.0     3.0       6.0    1.0     1.0    3-6-1-1       South   
1       50.0     3.0       6.0    1.0     3.0    3-6-1-3       South   
2       50.0     3.0       6.0    1.0     5.0    3-6-1-5       South   
3       50.0     3.0       6.0    1.0     7.0    3-6-1-7       South   
4       50.0     3.0       6.0    1.0     9.0    3-6-1-9       South   
...      ...     ...       ...    ...     ...        ...         ...   
3141    50.0     4.0       8.0   56.0    37.0  4-8-56-37        West   
3142    50.0     4.0       8.0   56.0    39.0  4-8-56-39        West   
3143    50.0     4.0       8.0   56.0    41.0  4-8-56-41        West   
3144    50.0     4.0       8.0   56.0    43.0  4-8-56-43        West   
3145    50.0     4.0       8.0   56.0    45.0  4-8-56-45        West   

           division_name  
0     East South Central  
1     East South Central  
2     East South Central  
3     East South Central  
4     East South Central  
...                  ...  
3141            Mountain  
3142            Mountain  
3143            Mountain  
3144            Mountain  
3145            Mountain  

[3146 rows x 13 columns]

In [30]:
new= pd.merge(left=age, right=master, on="fips",how='right') 
# merge two dataframe and based on the fips index in the both dataframe and also make sure there is one row for 
# every county in the new dataframe from county_fips_master
new.shape #get the dimension of the dataframe
new

fips     0-17    18-24    25-34    35-44    45-54    55-64    65-74  \
0     01001  25941.0  11422.0  12315.0  13828.0  14000.0  12697.0   9594.0   
1     01003  86587.0  37568.0  44133.0  46730.0  49675.0  52405.0  43252.0   
2     01005  11057.0   6162.0   6603.0   5907.0   6490.0   6377.0   5255.0   
3     01007   9671.0   5241.0   5788.0   5472.0   6707.0   5563.0   4270.0   
4     01009  25671.0  11360.0  12635.0  13570.0  14737.0  14123.0  12106.0   
...     ...      ...      ...      ...      ...      ...      ...      ...   
3141  56043   3847.0   1441.0   1815.0   1804.0   2020.0   2350.0   1834.0   
3142  56045   2901.0   1336.0   1402.0   1719.0   1551.0   2118.0   1309.0   
3143  02270      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
3144  46113      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
3145  51515      NaN      NaN      NaN      NaN      NaN      NaN      NaN   

        75-84     85+  ...    state_name                    long_name  sumlev  \
0      5430.0  1945.0  ...       Alabama            Autauga County AL    50.0   
1     23262.0  8854.0  ...       Alabama            Baldwin County AL    50.0   
2      2795.0  1074.0  ...       Alabama            Barbour County AL    50.0   
3      2555.0   638.0  ...       Alabama               Bibb County AL    50.0   
4      6560.0  2022.0  ...       Alabama             Blount County AL    50.0   
...       ...     ...  ...           ...                          ...     ...   
3141   1153.0   517.0  ...       Wyoming           Washakie County WY    50.0   
3142    835.0   469.0  ...       Wyoming             Weston County WY    50.0   
3143      NaN     NaN  ...        Alaska  Wade Hampton Census Area AK     NaN   
3144      NaN     NaN  ...  South Dakota            Shannon County SD     NaN   
3145      NaN     NaN  ...      Virginia              Bedford city VA     NaN   

      region  division  state  county    crosswalk  region_name  \
0        3.0       6.0    1.0     1.0      3-6-1-1        South   
1        3.0       6.0    1.0     3.0      3-6-1-3        South   
2        3.0       6.0    1.0     5.0      3-6-1-5        South   
3        3.0       6.0    1.0     7.0      3-6-1-7        South   
4        3.0       6.0    1.0     9.0      3-6-1-9        South   
...      ...       ...    ...     ...          ...          ...   
3141     4.0       8.0   56.0    43.0    4-8-56-43         West   
3142     4.0       8.0   56.0    45.0    4-8-56-45         West   
3143     NaN       NaN    NaN     NaN  NA-NA-NA-NA          NaN   
3144     NaN       NaN    NaN     NaN  NA-NA-NA-NA          NaN   
3145     NaN       NaN    NaN     NaN  NA-NA-NA-NA          NaN   

           division_name  
0     East South Central  
1     East South Central  
2     East South Central  
3     East South Central  
4     East South Central  
...                  ...  
3141            Mountain  
3142            Mountain  
3143                 NaN  
3144                 NaN  
3145                 NaN  

[3146 rows x 33 columns]

**There are 3146 rows in the new dataframe.**

In [31]:
#new[new.isna().any(axis=1)] #select the row with NA 
new[new.isnull().any(axis=1)]

fips  0-17  18-24  25-34  35-44  45-54  55-64  65-74  75-84  85+  ...  \
3143  02270   NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  NaN  ...   
3144  46113   NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  NaN  ...   
3145  51515   NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  NaN  ...   

        state_name                    long_name  sumlev  region  division  \
3143        Alaska  Wade Hampton Census Area AK     NaN     NaN       NaN   
3144  South Dakota            Shannon County SD     NaN     NaN       NaN   
3145      Virginia              Bedford city VA     NaN     NaN       NaN   

      state  county    crosswalk  region_name  division_name  
3143    NaN     NaN  NA-NA-NA-NA          NaN            NaN  
3144    NaN     NaN  NA-NA-NA-NA          NaN            NaN  
3145    NaN     NaN  NA-NA-NA-NA          NaN            NaN  

[3 rows x 33 columns]

In [32]:
new[new.isna().any(axis=1)].shape #get the dimension of the selected dateframe which contains NA

(3, 33)

**There are three rows have missingness.**

__Exercise 3.2 (10 points)__ Drop any rows with missingness.  For each state calculate the proportion of each age bin as you did for the entire US in the question above.  What is the state with the most 0-17 proportion and the state with the most 85+ proportion?  Give the proportions.

In [47]:
new_1= new.dropna() #remove the row with miss values.
new_2= new_1.groupby("state_name").sum() #grouping the data based on state
new_3= new_2.iloc[:, 0:9] #select the data which has population and number of people in each age bin
new_3.head()

0-17      18-24       25-34      35-44      45-54  \
state_name                                                            
Alabama      2168219.0  1135562.0   1197337.0  1141813.0  1190723.0   
Alaska        331933.0   170067.0    201417.0   163690.0   173534.0   
Arizona      3050038.0  1566538.0   1658537.0  1530872.0  1530958.0   
Arkansas     1380988.0   688947.0    730870.0   696628.0   708892.0   
California  17087279.0  8967102.0  10257203.0  9249327.0  9236216.0   

                55-64      65-74      75-84        85+  
state_name                                              
Alabama     1220538.0   908803.0   502962.0   204651.0  
Alaska       175823.0   102980.0    44886.0    18826.0  
Arizona     1556376.0  1258193.0   703632.0   294345.0  
Arkansas     730473.0   565342.0   312872.0   132352.0  
California  8753500.0  6121938.0  3438959.0  1634514.0

In [48]:
new_4= new_3.div(new_2['population'],axis=0) #divide by single series instead of dataframe. column-wise calculation, axis=0; row-wise axis=1.

In [49]:
new_5= new_4.reset_index() #reset the index to make sure the first column is state.
print(new_5.sort_values(by=['0-17'],ascending=False).iloc[0]) 
#sort the data (large-small) based on the 0-17 age ratio and then output is the fips of the max
print(new_5.sort_values(by=['85+'],ascending=False).iloc[0])
#sort the data (large-small) based on the 85+ age ratio and then output is the fips of the max
#https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html .sort_value

state_name         Utah
0-17           0.282481
18-24          0.132711
25-34          0.139205
35-44          0.122167
45-54          0.100823
55-64         0.0986018
65-74         0.0689702
75-84         0.0387066
85+            0.016336
Name: 44, dtype: object
state_name    Rhode Island
0-17              0.202001
18-24             0.126987
25-34             0.124965
35-44             0.113423
45-54             0.128491
55-64             0.130042
65-74            0.0914052
75-84             0.052858
85+              0.0298275
Name: 39, dtype: object


**Utah has the most 0-17 proportion which is 0.282481; Rhode Island has most 85+ proportion which is 0.0298275.**

### Part 4. Read in the 'time_series_covid19_deaths_US.csv' data and answer the next five questions

__Exercise 4.1 (5 points).__ This data is the Johns Hopkins official US Covid-19 death cumulative counts for each date and county.  Restructure the dataframe such that it is 'tidy' and there is a column for date, deaths, population, and fips.  (Hint: look at pandas melt function)

In [51]:
df = pd.read_csv("time_series_covid19_deaths_US.csv")#load date
list= df.columns[12:413].tolist() # choose the column which contain dates
df_1= pd.melt(df, id_vars = ['FIPS','Population'], value_vars=list, var_name='Date', value_name='Death')
#change the structure of data
df_1 

FIPS  Population     Date  Death
0         1001.0       55869  1/22/20      0
1         1003.0      223234  1/22/20      0
2         1005.0       24686  1/22/20      0
3         1007.0       22394  1/22/20      0
4         1009.0       57826  1/22/20      0
...          ...         ...      ...    ...
1339335  56039.0       23464  2/25/21      9
1339336  56041.0       20226  2/25/21     12
1339337  90056.0           0  2/25/21      0
1339338  56043.0        7805  2/25/21     26
1339339  56045.0        6927  2/25/21      5

[1339340 rows x 4 columns]

__Exercise 4.2 (5 points).__ Which county (give county name and state) has the highest proportion of deaths by 2/25/21 to the Population?  Use the population in the covid deaths datasets.

In [52]:
df['Proportion'] = df['2/25/21'] / df['Population'] #calculate the propotion of death by 2/25/21
df

UID iso2 iso3  code3     FIPS      Admin2 Province_State  \
0     84001001   US  USA    840   1001.0     Autauga        Alabama   
1     84001003   US  USA    840   1003.0     Baldwin        Alabama   
2     84001005   US  USA    840   1005.0     Barbour        Alabama   
3     84001007   US  USA    840   1007.0        Bibb        Alabama   
4     84001009   US  USA    840   1009.0      Blount        Alabama   
...        ...  ...  ...    ...      ...         ...            ...   
3335  84056039   US  USA    840  56039.0       Teton        Wyoming   
3336  84056041   US  USA    840  56041.0       Uinta        Wyoming   
3337  84090056   US  USA    840  90056.0  Unassigned        Wyoming   
3338  84056043   US  USA    840  56043.0    Washakie        Wyoming   
3339  84056045   US  USA    840  56045.0      Weston        Wyoming   

     Country_Region        Lat       Long_  ... 2/17/21  2/18/21  2/19/21  \
0                US  32.539527  -86.644082  ...      84       84       85   
1                US  30.727750  -87.722071  ...     253      254      262   
2                US  31.868263  -85.387129  ...      49       49       49   
3                US  32.996421  -87.125115  ...      57       57       58   
4                US  33.982109  -86.567906  ...     121      121      125   
...             ...        ...         ...  ...     ...      ...      ...   
3335             US  43.935225 -110.589080  ...       9        9        9   
3336             US  41.287818 -110.547578  ...      12       12       12   
3337             US   0.000000    0.000000  ...       0        0        0   
3338             US  43.904516 -107.680187  ...      26       26       26   
3339             US  43.839612 -104.567488  ...       5        5        5   

      2/20/21  2/21/21  2/22/21  2/23/21  2/24/21  2/25/21  Proportion  
0          85       85       85       84       85       89    0.001593  
1         262      262      262      263      272      274    0.001227  
2          50       50       50       50       50       51    0.002066  
3          58       58       58       59       59       60    0.002679  
4         125      125      125      125      125      125    0.002162  
...       ...      ...      ...      ...      ...      ...         ...  
3335        9        9        9        9        9        9    0.000384  
3336       12       12       12       12       12       12    0.000593  
3337        0        0        0        0        0        0         NaN  
3338       26       26       26       26       26       26    0.003331  
3339        5        5        5        5        5        5    0.000722  

[3340 rows x 414 columns]

In [53]:
# Replacing infinite with nan
df.replace([np.inf, -np.inf], np.nan, inplace=True)

# Dropping all the rows with nan values
df.dropna(inplace=True)

# Printing df
df
#https://www.geeksforgeeks.org/remove-infinite-values-from-a-given-pandas-dataframe/

UID iso2 iso3  code3     FIPS      Admin2 Province_State  \
0     84001001   US  USA    840   1001.0     Autauga        Alabama   
1     84001003   US  USA    840   1003.0     Baldwin        Alabama   
2     84001005   US  USA    840   1005.0     Barbour        Alabama   
3     84001007   US  USA    840   1007.0        Bibb        Alabama   
4     84001009   US  USA    840   1009.0      Blount        Alabama   
...        ...  ...  ...    ...      ...         ...            ...   
3334  84056037   US  USA    840  56037.0  Sweetwater        Wyoming   
3335  84056039   US  USA    840  56039.0       Teton        Wyoming   
3336  84056041   US  USA    840  56041.0       Uinta        Wyoming   
3338  84056043   US  USA    840  56043.0    Washakie        Wyoming   
3339  84056045   US  USA    840  56045.0      Weston        Wyoming   

     Country_Region        Lat       Long_  ... 2/17/21  2/18/21  2/19/21  \
0                US  32.539527  -86.644082  ...      84       84       85   
1                US  30.727750  -87.722071  ...     253      254      262   
2                US  31.868263  -85.387129  ...      49       49       49   
3                US  32.996421  -87.125115  ...      57       57       58   
4                US  33.982109  -86.567906  ...     121      121      125   
...             ...        ...         ...  ...     ...      ...      ...   
3334             US  41.659439 -108.882788  ...      34       34       34   
3335             US  43.935225 -110.589080  ...       9        9        9   
3336             US  41.287818 -110.547578  ...      12       12       12   
3338             US  43.904516 -107.680187  ...      26       26       26   
3339             US  43.839612 -104.567488  ...       5        5        5   

      2/20/21  2/21/21  2/22/21  2/23/21  2/24/21  2/25/21  Proportion  
0          85       85       85       84       85       89    0.001593  
1         262      262      262      263      272      274    0.001227  
2          50       50       50       50       50       51    0.002066  
3          58       58       58       59       59       60    0.002679  
4         125      125      125      125      125      125    0.002162  
...       ...      ...      ...      ...      ...      ...         ...  
3334       34       34       34       34       34       34    0.000803  
3335        9        9        9        9        9        9    0.000384  
3336       12       12       12       12       12       12    0.000593  
3338       26       26       26       26       26       26    0.003331  
3339        5        5        5        5        5        5    0.000722  

[3220 rows x 414 columns]

In [54]:
a= df.sort_values(by=['Proportion'],ascending=False) #sort the data based on the value of proportion 
a.head()

UID iso2 iso3  code3     FIPS   Admin2 Province_State  \
2568  84046073   US  USA    840  46073.0  Jerauld   South Dakota   
955   84020063   US  USA    840  20063.0     Gove         Kansas   
3043  84051595   US  USA    840  51595.0  Emporia       Virginia   
3055  84051640   US  USA    840  51640.0    Galax       Virginia   
2775  84048155   US  USA    840  48155.0    Foard          Texas   

     Country_Region        Lat       Long_  ... 2/17/21  2/18/21  2/19/21  \
2568             US  44.066378  -98.629660  ...      16       16       16   
955              US  38.915882 -100.482932  ...      19       19       19   
3043             US  36.694556  -77.534061  ...      35       35       35   
3055             US  36.666019  -80.917031  ...      43       43       43   
2775             US  33.973086  -99.776780  ...       7        7        8   

      2/20/21  2/21/21  2/22/21  2/23/21  2/24/21  2/25/21  Proportion  
2568       16       16       16       16       16       16    0.007948  
955        19       19       19       19       19       19    0.007208  
3043       35       35       37       37       38       38    0.007108  
3055       43       43       44       44       44       44    0.006932  
2775        8        8        8        8        8        8    0.006926  

[5 rows x 414 columns]

**Jerauld in South Dakota state has the highest proportion of deaths (0.007948) by 2/25/21 to the Population.**

__Exercise 4.3 (10 points).__ Compute the incidence proportion (number of deaths each day per 100,000 population) for each county, you may save this as a new dataset.

In [55]:
df_2 = df[df.columns[12:413]].diff(axis=1) #calculate the new death for each day
df_2
#diff function found at https://pythontic.com/pandas/dataframe-computations/difference

1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  1/28/20  1/29/20  \
0         NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
1         NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
2         NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
3         NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
4         NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
...       ...      ...      ...      ...      ...      ...      ...      ...   
3334      NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
3335      NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
3336      NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
3338      NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
3339      NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   

      1/30/20  1/31/20  ...  2/16/21  2/17/21  2/18/21  2/19/21  2/20/21  \
0         0.0      0.0  ...      0.0      0.0      0.0      1.0      0.0   
1         0.0      0.0  ...      1.0      0.0      1.0      8.0      0.0   
2         0.0      0.0  ...      0.0      1.0      0.0      0.0      1.0   
3         0.0      0.0  ...      0.0      0.0      0.0      1.0      0.0   
4         0.0      0.0  ...      0.0      0.0      0.0      4.0      0.0   
...       ...      ...  ...      ...      ...      ...      ...      ...   
3334      0.0      0.0  ...      0.0      0.0      0.0      0.0      0.0   
3335      0.0      0.0  ...      1.0      0.0      0.0      0.0      0.0   
3336      0.0      0.0  ...      0.0      0.0      0.0      0.0      0.0   
3338      0.0      0.0  ...      0.0      0.0      0.0      0.0      0.0   
3339      0.0      0.0  ...      0.0      0.0      0.0      0.0      0.0   

      2/21/21  2/22/21  2/23/21  2/24/21  2/25/21  
0         0.0      0.0     -1.0      1.0      4.0  
1         0.0      0.0      1.0      9.0      2.0  
2         0.0      0.0      0.0      0.0      1.0  
3         0.0      0.0      1.0      0.0      1.0  
4         0.0      0.0      0.0      0.0      0.0  
...       ...      ...      ...      ...      ...  
3334      0.0      0.0      0.0      0.0      0.0  
3335      0.0      0.0      0.0      0.0      0.0  
3336      0.0      0.0      0.0      0.0      0.0  
3338      0.0      0.0      0.0      0.0      0.0  
3339      0.0      0.0      0.0      0.0      0.0  

[3220 rows x 401 columns]

In [56]:
df_3=df_2.div(df['Population'],axis=0)*100000 #calculate the incidence proportion
df_3

1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  1/28/20  1/29/20  \
0         NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
1         NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
2         NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
3         NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
4         NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
...       ...      ...      ...      ...      ...      ...      ...      ...   
3334      NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
3335      NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
3336      NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
3338      NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
3339      NaN      0.0      0.0      0.0      0.0      0.0      0.0      0.0   

      1/30/20  1/31/20  ...   2/16/21   2/17/21  2/18/21   2/19/21   2/20/21  \
0         0.0      0.0  ...  0.000000  0.000000  0.00000  1.789901  0.000000   
1         0.0      0.0  ...  0.447960  0.000000  0.44796  3.583683  0.000000   
2         0.0      0.0  ...  0.000000  4.050879  0.00000  0.000000  4.050879   
3         0.0      0.0  ...  0.000000  0.000000  0.00000  4.465482  0.000000   
4         0.0      0.0  ...  0.000000  0.000000  0.00000  6.917304  0.000000   
...       ...      ...  ...       ...       ...      ...       ...       ...   
3334      0.0      0.0  ...  0.000000  0.000000  0.00000  0.000000  0.000000   
3335      0.0      0.0  ...  4.261848  0.000000  0.00000  0.000000  0.000000   
3336      0.0      0.0  ...  0.000000  0.000000  0.00000  0.000000  0.000000   
3338      0.0      0.0  ...  0.000000  0.000000  0.00000  0.000000  0.000000   
3339      0.0      0.0  ...  0.000000  0.000000  0.00000  0.000000  0.000000   

      2/21/21  2/22/21   2/23/21   2/24/21   2/25/21  
0         0.0      0.0 -1.789901  1.789901  7.159606  
1         0.0      0.0  0.447960  4.031644  0.895921  
2         0.0      0.0  0.000000  0.000000  4.050879  
3         0.0      0.0  4.465482  0.000000  4.465482  
4         0.0      0.0  0.000000  0.000000  0.000000  
...       ...      ...       ...       ...       ...  
3334      0.0      0.0  0.000000  0.000000  0.000000  
3335      0.0      0.0  0.000000  0.000000  0.000000  
3336      0.0      0.0  0.000000  0.000000  0.000000  
3338      0.0      0.0  0.000000  0.000000  0.000000  
3339      0.0      0.0  0.000000  0.000000  0.000000  

[3220 rows x 401 columns]

In [57]:
df_4= df.loc[:, ('FIPS','Admin2','Province_State')] #select the fips, state and county information
df_5= pd.concat([df_4,df_3,],axis=1) #merge and finalize the table
df_5.head()

FIPS   Admin2 Province_State  1/22/20  1/23/20  1/24/20  1/25/20  \
0  1001.0  Autauga        Alabama      NaN      0.0      0.0      0.0   
1  1003.0  Baldwin        Alabama      NaN      0.0      0.0      0.0   
2  1005.0  Barbour        Alabama      NaN      0.0      0.0      0.0   
3  1007.0     Bibb        Alabama      NaN      0.0      0.0      0.0   
4  1009.0   Blount        Alabama      NaN      0.0      0.0      0.0   

   1/26/20  1/27/20  1/28/20  ...  2/16/21   2/17/21  2/18/21   2/19/21  \
0      0.0      0.0      0.0  ...  0.00000  0.000000  0.00000  1.789901   
1      0.0      0.0      0.0  ...  0.44796  0.000000  0.44796  3.583683   
2      0.0      0.0      0.0  ...  0.00000  4.050879  0.00000  0.000000   
3      0.0      0.0      0.0  ...  0.00000  0.000000  0.00000  4.465482   
4      0.0      0.0      0.0  ...  0.00000  0.000000  0.00000  6.917304   

    2/20/21  2/21/21  2/22/21   2/23/21   2/24/21   2/25/21  
0  0.000000      0.0      0.0 -1.789901  1.789901  7.159606  
1  0.000000      0.0      0.0  0.447960  4.031644  0.895921  
2  4.050879      0.0      0.0  0.000000  0.000000  4.050879  
3  0.000000      0.0      0.0  4.465482  0.000000  4.465482  
4  0.000000      0.0      0.0  0.000000  0.000000  0.000000  

[5 rows x 404 columns]

__Exercise 4.4 (15 points).__ Compute the incidence proportion per week (per 100,000 population), which is the sum of the incidences for all days in a week.  Hint: use Periods in Pandas.  Which is the week with the highest mean incidence proportions (where the mean is taken over all the counties)?

In [58]:
list= df_5.columns[3:].tolist() # choose the column which contain dates
df_6= pd.melt(df_5, id_vars = ['Admin2'], value_vars= list,var_name='Date', value_name='Death')
df_6

Admin2     Date  Death
0           Autauga  1/22/20    NaN
1           Baldwin  1/22/20    NaN
2           Barbour  1/22/20    NaN
3              Bibb  1/22/20    NaN
4            Blount  1/22/20    NaN
...             ...      ...    ...
1291215  Sweetwater  2/25/21    0.0
1291216       Teton  2/25/21    0.0
1291217       Uinta  2/25/21    0.0
1291218    Washakie  2/25/21    0.0
1291219      Weston  2/25/21    0.0

[1291220 rows x 3 columns]

In [59]:
df_6['Date'] = pd.to_datetime(df_6['Date']) #change Date type to datetime
# https://pandas.pydata.org/docs/reference/api/pandas.PeriodIndex.html.

In [60]:
import datetime #https://www.tutorialspoint.com/How-can-I-get-the-current-week-using-Python#:~:text=YOu%20can%20use%20the%20isocalender,number%20and%20Day%20of%20Week.
#write a function to generate week information from a date
def week_func(my_date):
    my_date = pd.to_datetime(my_date) #change type
    year, week_num, day_of_week = my_date.isocalendar() #extract week number and year
    i= ("Week #" + str(week_num) + " of year " + str(year))
    return (i)

In [61]:
df_6['Week'] = df_6['Date'].apply(week_func) #apply week_func on column 'date' 
display(df_6.tail())
df_7 = df_6.groupby('Week').sum() #group the data and sum them based on Week
display(df_7)
df_8= df_6.groupby('Week').mean() #group the data and calculate the mean based on Week
df_8.idxmax(axis = 0) #find the maximum 

Admin2       Date  Death                  Week
1291215  Sweetwater 2021-02-25    0.0  Week #8 of year 2021
1291216       Teton 2021-02-25    0.0  Week #8 of year 2021
1291217       Uinta 2021-02-25    0.0  Week #8 of year 2021
1291218    Washakie 2021-02-25    0.0  Week #8 of year 2021
1291219      Weston 2021-02-25    0.0  Week #8 of year 2021

Death
Week                               
Week #1 of year 2021   27273.269829
Week #10 of year 2020      1.755252
Week #11 of year 2020      7.538573
Week #12 of year 2020    233.190633
Week #13 of year 2020    698.865353
Week #14 of year 2020   2253.088550
Week #15 of year 2020   3823.740834
Week #16 of year 2020   4728.229116
Week #17 of year 2020   5504.047455
Week #18 of year 2020   5161.336782
Week #19 of year 2020   4944.563461
Week #2 of year 2021   26767.680038
Week #20 of year 2020   4565.006800
Week #21 of year 2020   4568.135354
Week #22 of year 2020   3963.892473
Week #23 of year 2020   3501.499107
Week #24 of year 2020   3325.849460
Week #25 of year 2020   2564.202523
Week #26 of year 2020   3434.084186
Week #27 of year 2020   2775.283861
Week #28 of year 2020   3581.771549
Week #29 of year 2020   3760.553204
Week #3 of year 2021   23760.755305
Week #30 of year 2020   5279.704075
Week #31 of year 2020   7536.804154
Week #32 of year 2020   7043.776239
Week #33 of year 2020   7062.211933
Week #34 of year 2020   7544.980105
Week #35 of year 2020   7384.626639
Week #36 of year 2020   6855.860239
Week #37 of year 2020   6280.849024
Week #38 of year 2020   7411.374076
Week #39 of year 2020   7496.583448
Week #4 of year 2020       0.000000
Week #4 of year 2021   24503.498123
Week #40 of year 2020   7579.808545
Week #41 of year 2020   7887.695636
Week #42 of year 2020   9089.269882
Week #43 of year 2020  11255.924374
Week #44 of year 2020  11930.257463
Week #45 of year 2020  14088.862766
Week #46 of year 2020  15086.903552
Week #47 of year 2020  19459.326925
Week #48 of year 2020  18387.483878
Week #49 of year 2020  26397.917045
Week #5 of year 2020       0.000000
Week #5 of year 2021   24694.788140
Week #50 of year 2020  28950.157434
Week #51 of year 2020  28883.608820
Week #52 of year 2020  20121.439213
Week #53 of year 2020  23604.737546
Week #6 of year 2020       0.000000
Week #6 of year 2021   17819.587131
Week #7 of year 2020       0.000000
Week #7 of year 2021   13042.162498
Week #8 of year 2020       0.000000
Week #8 of year 2021    9345.965637
Week #9 of year 2020       0.044390

Death    Week #50 of year 2020
dtype: object

**Table7 shows the sum of the incidences for all days in a week.**

**Week #50 of year 2020 (Dec. 7, 2020	Dec. 13, 2020) has the highest mean incidence proportions.**

__Exercise 4.5 (15 points).__ For each week (should be roughly 40 weeks), compute the spearman correlation between the proportion of seniors in the county and the incidence proportion for that county at that week (each county is considered to be a sample for computing the correlation).  Print all correlations below.  This should be a single number for each week.

In [268]:
merge_1= age.loc[:,('fips','population','65-74','75-84', '85+')] #find aim is to find the sum of senior
merge_1.rename(columns={'fips': 'FIPS'},inplace=True)
merge_1 #change the structure of the dataframe

FIPS  population  65-74  75-84   85+
0     01001      107172   9594   5430  1945
1     01003      392466  43252  23262  8854
2     01005       51720   5255   2795  1074
3     01007       45905   4270   2555   638
4     01009      112784  12106   6560  2022
...     ...         ...    ...    ...   ...
3215  72145      121223  11928   6139  3159
3216  72147       21708   2364   1498   479
3217  72149       48348   4631   2218   859
3218  72151       75923   7926   3784  2103
3219  72153       81465   8571   4620  2346

[3220 rows x 5 columns]

In [269]:
list= df_5.columns[3:].tolist() # choose the column which contain dates
df_7= pd.melt(df_5, id_vars = ['FIPS','Admin2'], value_vars= list,var_name='Date', value_name='Death')
#########
df_7.replace([np.inf, -np.inf], np.nan, inplace=True)
# Dropping all the rows with nan values
df_7.dropna(inplace=True)

###########
df_7['Date'] = pd.to_datetime(df_7['Date']) #change date type
df_7.set_index('Date', inplace = True) #set date as index
df_7 = df_7.astype({"FIPS": int}).astype({"FIPS": str}) # change fips to string

df_7['FIPS']= df_7['FIPS'].str.zfill(5) #make the table consistant
df_7 = df_7.reset_index().set_index('FIPS') #set fips as index
df_7.head(n=100)


Date         Admin2  Death
FIPS                                  
01001 2020-01-23        Autauga    0.0
01003 2020-01-23        Baldwin    0.0
01005 2020-01-23        Barbour    0.0
01007 2020-01-23           Bibb    0.0
01009 2020-01-23         Blount    0.0
...          ...            ...    ...
02290 2020-01-23  Yukon-Koyukuk    0.0
04001 2020-01-23         Apache    0.0
04003 2020-01-23        Cochise    0.0
04005 2020-01-23       Coconino    0.0
04007 2020-01-23           Gila    0.0

[100 rows x 3 columns]

In [275]:
fips_merge= pd.merge(left=df_7, right=merge_1, on="FIPS",how='left') #merge two dataset based on fips
fips_merge


FIPS       Date      Admin2  Death  population  65-74  75-84   85+
0        01001 2020-01-23     Autauga    0.0      107172   9594   5430  1945
1        01003 2020-01-23     Baldwin    0.0      392466  43252  23262  8854
2        01005 2020-01-23     Barbour    0.0       51720   5255   2795  1074
3        01007 2020-01-23        Bibb    0.0       45905   4270   2555   638
4        01009 2020-01-23      Blount    0.0      112784  12106   6560  2022
...        ...        ...         ...    ...         ...    ...    ...   ...
1287995  56037 2021-02-25  Sweetwater    0.0       90261   6216   3372  1352
1287996  56039 2021-02-25       Teton    0.0       48950   4164   1593   990
1287997  56041 2021-02-25       Uinta    0.0       44225   3544   1457   805
1287998  56043 2021-02-25    Washakie    0.0       16781   1834   1153   517
1287999  56045 2021-02-25      Weston    0.0       13640   1309    835   469

[1288000 rows x 8 columns]

In [277]:
fips_merge['senior_rat'] = (fips_merge["65-74"] + fips_merge["75-84"] + fips_merge["85+"])/fips_merge["population"] 
#calculate senior population
fips_merge= fips_merge.drop(['FIPS','65-74', '75-84','85+'], axis=1) #remove some columns which are not needed
fips_merge

Date      Admin2  Death  population  senior_rat
0       2020-01-23     Autauga    0.0      107172    0.158334
1       2020-01-23     Baldwin    0.0      392466    0.192037
2       2020-01-23     Barbour    0.0       51720    0.176411
3       2020-01-23        Bibb    0.0       45905    0.162575
4       2020-01-23      Blount    0.0      112784    0.183430
...            ...         ...    ...         ...         ...
1287995 2021-02-25  Sweetwater    0.0       90261    0.121204
1287996 2021-02-25       Teton    0.0       48950    0.137835
1287997 2021-02-25       Uinta    0.0       44225    0.131283
1287998 2021-02-25    Washakie    0.0       16781    0.208808
1287999 2021-02-25      Weston    0.0       13640    0.191569

[1288000 rows x 5 columns]

In [278]:
fips_merge['Week'] = fips_merge['Date'].apply(week_func) #apply week_func on Date and save as a new column Week.

In [297]:
fips_merge

Date      Admin2  Death  population  senior_rat  \
0       2020-01-23     Autauga    0.0      107172    0.158334   
1       2020-01-23     Baldwin    0.0      392466    0.192037   
2       2020-01-23     Barbour    0.0       51720    0.176411   
3       2020-01-23        Bibb    0.0       45905    0.162575   
4       2020-01-23      Blount    0.0      112784    0.183430   
...            ...         ...    ...         ...         ...   
1287995 2021-02-25  Sweetwater    0.0       90261    0.121204   
1287996 2021-02-25       Teton    0.0       48950    0.137835   
1287997 2021-02-25       Uinta    0.0       44225    0.131283   
1287998 2021-02-25    Washakie    0.0       16781    0.208808   
1287999 2021-02-25      Weston    0.0       13640    0.191569   

                         Week  
0        Week #4 of year 2020  
1        Week #4 of year 2020  
2        Week #4 of year 2020  
3        Week #4 of year 2020  
4        Week #4 of year 2020  
...                       ...  
1287995  Week #8 of year 2021  
1287996  Week #8 of year 2021  
1287997  Week #8 of year 2021  
1287998  Week #8 of year 2021  
1287999  Week #8 of year 2021  

[1288000 rows x 6 columns]

In [280]:
index = fips_merge.Week.unique()#find the unique week in the whole data
index

array(['Week #4 of year 2020', 'Week #5 of year 2020',
       'Week #6 of year 2020', 'Week #7 of year 2020',
       'Week #8 of year 2020', 'Week #9 of year 2020',
       'Week #10 of year 2020', 'Week #11 of year 2020',
       'Week #12 of year 2020', 'Week #13 of year 2020',
       'Week #14 of year 2020', 'Week #15 of year 2020',
       'Week #16 of year 2020', 'Week #17 of year 2020',
       'Week #18 of year 2020', 'Week #19 of year 2020',
       'Week #20 of year 2020', 'Week #21 of year 2020',
       'Week #22 of year 2020', 'Week #23 of year 2020',
       'Week #24 of year 2020', 'Week #25 of year 2020',
       'Week #26 of year 2020', 'Week #27 of year 2020',
       'Week #28 of year 2020', 'Week #29 of year 2020',
       'Week #30 of year 2020', 'Week #31 of year 2020',
       'Week #32 of year 2020', 'Week #33 of year 2020',
       'Week #34 of year 2020', 'Week #35 of year 2020',
       'Week #36 of year 2020', 'Week #37 of year 2020',
       'Week #38 of year 2020', 'Week

In [287]:
from scipy.stats import spearmanr   
#write a for loop function to calculate spearman correlation
for i in index:
    a= fips_merge.loc[fips_merge["Week"]==i,:] #for all unique weeks
    cor  = spearmanr(a['Death'], a['senior_rat']) #select death and senior ratio and calculate correlation
    print (cor) #print each of them


SpearmanrResult(correlation=nan, pvalue=nan)
SpearmanrResult(correlation=nan, pvalue=nan)
SpearmanrResult(correlation=nan, pvalue=nan)
SpearmanrResult(correlation=nan, pvalue=nan)
SpearmanrResult(correlation=nan, pvalue=nan)
SpearmanrResult(correlation=-0.008853402323476943, pvalue=0.18380092496328648)
SpearmanrResult(correlation=-0.01987281043582171, pvalue=0.002847865402744172)
SpearmanrResult(correlation=-0.03948446653776761, pvalue=3.030901280098327e-09)
SpearmanrResult(correlation=-0.08829774191737347, pvalue=2.9584028145115867e-40)
SpearmanrResult(correlation=-0.1399938524926244, pvalue=5.138296628378682e-99)
SpearmanrResult(correlation=-0.18240917603040138, pvalue=7.118608730619206e-168)
SpearmanrResult(correlation=-0.19555129502830468, pvalue=4.08247291176184e-193)
SpearmanrResult(correlation=-0.20164134193084368, pvalue=1.8752853796570944e-205)
SpearmanrResult(correlation=-0.20770214338748488, pvalue=3.8664025113087664e-218)
SpearmanrResult(correlation=-0.21781189818544164, pv